FEXCore/Source/Interface/IR/Passes/RegisterAllocationPass.cpp

```diff
@@ -5,1239 +5,713 @@ tags: ir|opts
-  Graph->BlockPredecessors.clear();
+        SourceIndex--;
+        LOGMAN_THROW_AA_FMT(SourceIndex >= 0, "Consistent source count");
 
-  for (auto [BlockNode, BlockIROp] : IR->GetBlocks()) {
-    auto CodeBlock = BlockIROp->C<IROp_CodeBlock>();
+        OrderedNode* Old = IR->GetNode(IROp->Args[s]);
+        LOGMAN_THROW_AA_FMT(IsInRegisterFile(Old), "sources in file");
 
-    auto IROp = IR->GetNode(IR->GetNode(CodeBlock->Last)->Header.Previous)->Op(IR->GetData());
-    if (IROp->Op == OP_JUMP) {
-      auto Op = IROp->C<IROp_Jump>();
-      Graph->BlockPredecessors[Op->TargetBlock.ID()].insert(IR->GetID(BlockNode));
-    } else if (IROp->Op == OP_CONDJUMP) {
-      auto Op = IROp->C<IROp_CondJump>();
-      Graph->BlockPredecessors[Op->TrueBlock.ID()].insert(IR->GetID(BlockNode));
-      Graph->BlockPredecessors[Op->FalseBlock.ID()].insert(IR->GetID(BlockNode));
-    }
-  }
-}
-
-bool ConstrainedRAPass::Run(IREmitter* IREmit) {
-  FEXCORE_PROFILE_SCOPED("PassManager::RA");
-  bool Changed = false;
-
-  auto IR = IREmit->ViewIR();
+        if (!SourcesNextUses[SourceIndex]) {
+          FreeReg(SSAToReg[IR->GetID(Map(Old)).Value]);
+        }
 
-  auto HeaderOp = IR.GetHeader();
-  OriginalRIP = HeaderOp->OriginalRIP;
-  SpillSlotCount = 0;
-  Graph->SpillStack.clear();
+        NextUses[IR->GetID(Old).Value] = SourcesNextUses[SourceIndex];
+      }
 
-  CalculatePredecessors(&IR);
+      // Assign destinations.
+      if (GetHasDest(IROp->Op)) {
+        AssignReg(IROp, CodeNode, IROp);
+      }
 
-  while (1) {
-    HadFullRA = true;
+      // Remap sources last, since AssignReg can shuffle.
+      for (auto s = 0; s < IR::GetRAArgs(IROp->Op); ++s) {
+        OrderedNode* Remapped = SSAToNewSSA[IR->GetID(IR->GetNode(IROp->Args[s])).Value];
 
-    // Virtual allocation pass runs the compaction pass per run
-    Changed |= RunAllocateVirtualRegisters(IREmit);
+        if (Remapped != nullptr) {
+          IREmit->ReplaceNodeArgument(CodeNode, s, Remapped);
+        }
+      }
 
-    if (HadFullRA) {
-      break;
+      LOGMAN_THROW_AA_FMT(IP >= 1, "IP relative to end of block, iterating forward");
+      --IP;
     }
 
-    SpillOne(IREmit);
-    Changed = true;
-    // We need to rerun compaction after spilling
-    CompactionPass->Run(IREmit);
+    LOGMAN_THROW_AA_FMT(SourceIndex == 0, "Consistent source count in block");
   }
 
-  Graph->AllocData->SpillSlotCount = Graph->SpillStack.size();
-
-  return Changed;
+  /* Now that we're done growing things, we can finalize our results.
+   *
+   * TODO: Rework RegisterAllocationData to remove this memcpy, it's pointless.
+   */
+  AllocData = RegisterAllocationData::Create(SSAToReg.size());
+  AllocData->SpillSlotCount = SpillSlotCount;
+  memcpy(AllocData->Map, SSAToReg.data(), sizeof(PhysicalRegister) * SSAToReg.size());
+
+  PreferredReg.clear();
+  SSAToNewSSA.clear();
+  NewSSAToSSA.clear();
+  SSAToReg.clear();
+  SpillSlots.clear();
+  NextUses.clear();
```

Comment:
I would like to bulldoze the whole pass framework at some point because persisting these allocations is terrifying. In the mean time, this is probably fine.